pkg/cert/v1beta1/generate/generate.go

```diff
@@ -0,0 +1,262 @@
+		BasicConstraintsValid: false,
+	}
+
+	var err error
+	keyPair.key, err = rsa.GenerateKey(rand.Reader, 2048)
+	if err != nil {
+		return err
+	}
+	der, err := x509.CreateCertificate(rand.Reader, template, caKeyPair.cert, keyPair.key.Public(), caKeyPair.key)
+	if err != nil {
+		return err
+	}
+	if err = keyPair.encode(x509.MarshalPKCS1PrivateKey(keyPair.key), der); err != nil {
+		return err
+	}
+	klog.Info("Generate public certificate and private key signed with self-singed CA cert and private key.")
+	return nil
+}
+
+// createWebhookCertSecret create Secret embedded ca.key, ca.crt, tls.key and tls.cert
+func (o *generateOptions) createWebhookCertSecret(ctx context.Context, kubeClient client.Client, caKeyPair *certificates, keyPair *certificates) error {
+
+	certGeneratorJob := &batchv1.Job{}
+	if err := kubeClient.Get(ctx, client.ObjectKey{Namespace: o.namespace, Name: consts.JobName}, certGeneratorJob); err != nil {
+		return err
+	}
+
+	// Create secret with CA cert and server cert/key.
+	// Add ownerReferences to clean-up secret with cert generator Job.
+	jobUID := certGeneratorJob.UID
+	webhookCertSecret := &corev1.Secret{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "Secret",
+			APIVersion: "v1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      consts.Secret,
+			Namespace: o.namespace,
+			OwnerReferences: []metav1.OwnerReference{
+				{
+					APIVersion: "batch/v1",
+					Kind:       "Job",
+					Controller: pointer.BoolPtr(true),
+					Name:       consts.JobName,
+					UID:        jobUID,
+				},
+			},
+		},
+		Type: corev1.SecretTypeTLS,
+		Data: map[string][]byte{
+			"ca.key":  caKeyPair.keyPem,
+			"ca.crt":  caKeyPair.certPem,
+			"tls.key": keyPair.keyPem,
+			"tls.crt": keyPair.certPem,
+		},
+	}
+
+	oldSecret := &corev1.Secret{}
+	err := kubeClient.Get(ctx, client.ObjectKey{Namespace: o.namespace, Name: consts.Secret}, oldSecret)
+	switch {
+	case err != nil && !k8serrors.IsNotFound(err):
+		return err
+	case err == nil:
+		klog.Warning("Previous secret was found and removed.")
+		if err = kubeClient.Delete(ctx, oldSecret); err != nil {
+			return err
+		}
+	}
+
+	klog.Infof("Creating Secret: %s", consts.Secret)
+	if err = kubeClient.Create(ctx, webhookCertSecret); err != nil {
+		return err
+	}
+	return nil
+}
+
+// injectCert apply patch to ValidatingWebhookConfiguration and MutatingWebhookConfiguration
+func (o *generateOptions) injectCert(ctx context.Context, kubeClient client.Client, caKeypair *certificates) error {
+	validatingConf := &admissionregistrationv1.ValidatingWebhookConfiguration{}
+	if err := kubeClient.Get(ctx, client.ObjectKey{Name: consts.Webhook}, validatingConf); err != nil {
+		return err
+	}
+	newValidatingConf := validatingConf.DeepCopy()
+	newValidatingConf.Webhooks[0].ClientConfig.CABundle = caKeypair.certPem
+
+	klog.Info("Trying to patch ValidatingWebhookConfiguration adding the caBundle.")
+	for i := 0; i <= consts.MaxTryPatchCount; i++ {
```

Comment:
What is the use-case when we can't patch it from the first time ?
Does that happen when cert-generator job starts before `ValidatingWebhookConfiguration` was created ?

Reply:
> > Does that happen when cert-generator job starts before ValidatingWebhookConfiguration was created ?

I implemented the same as previously.

https://github.com/kubeflow/katib/blob/master/hack/cert-generator.sh#L162

Now that I think about it, once is enough. I'd like to fix it.